### ISSUE TITLE
Major source code clean up and refactoring

### ISSUE THREAD

**Issue by alexwkleung** [alexwkleung/Iris#32]
I want to do a major source code clean up and refactoring when time permits. At the moment, the source code is very fragile and quite a mess due to the hack-y stuff I've done. While I surprisingly ended up with something usable, I realized the limitations and problems when moving towards future development of the app.

On the one hand, the back-end seems somewhat robust, so it's just mostly tweaking how I handle panics on the Rust side to be used in the front-end (refer to #23). Of course, system-level panics aren't always the option because an app crash may cause the user to lose data, and that's not ideal nor a good UX when there's no visible warning within the app. I might be missing a few things here, but overall the system-level stuff can be improved if needed.

On the other hand, the front-end is what I am most concerned about. This is where everything is tied in place and is really what makes or break the app. While some of the smaller portions of code are fine, the main logic is where I cut a lot of corners. You might think that the app feels pretty good at its current stage and I do agree with that. However behind that is a source code nightmare and it's simply not feasible to add more features with what I have right now. Thus, a major source code clean up and refactoring must be done before proceeding further. The front-end of Iris is extremely crucial to the user experience, so by fixing the core, the app should feel more robust and rigid, along with the source code being more flexible.

For example, event listeners are not disposed at all within the app's lifetime, so memory issues are a problem from my testing. The implementation to add the dispose event is seen in #20. Not only memory issues, but conflicting event listeners (i.e., mouse or keyboard related) can break the user experience and cause headaches in the source code. At the time I didn't know how to handle multiple event listeners the proper way (when it's necessary) and it forced me to create a bunch of hacks since it was my first time dealing with it.

As stated previously, this will happen when time permits. I am not sure when exactly, but I do want to be clear here. No new features will be added until this is completed. There may be a few minor updates prior, but it won't be related to the core.

I ultimately want to give room for actual future development, so this is why I want to emphasize this early on, even if it won't happen right away. The app is still in its early stages, so hack-y code and poor code design can absolutely happen and I'm not ashamed of it because it allows room for improvement.

**Comment by alexwkleung** [comment]
Progress for this will be in the [origin/core-cleanup](https://github.com/alexwkleung/Iris/tree/core-cleanup) branch. Once completed, it will be merged into the main branch.

The goal is to keep everything intact and as close to the current state of the app, but making the core more robust and stable.

**Comment by alexwkleung** [comment]
`origin/core-cleanup` won't include UI changes unless absolutely necessary. The current UI needs a bit of tweaking but overall it's not as problematic compared to the core. However despite being very messy internally, I will add the UI clean up in the [v0.2.0 tracker](https://github.com/alexwkleung/Iris/issues/2) to be done at a later time.

To be clear on this previous segment here:

> On the other hand, the front-end is what I am most concerned about....
> ... 

What I mean (in simple terms) is the user interaction related aspects and the code binding to the UI elements; not the actual UI visuals.
